File: HackerRank/stf.py
from math import *
from fractions import *

parent = []
rank = []
edges = []# tam, num, denom, s, t

def find(x):
	if parent[x] != x:
		parent[x] = find(parent[x])
	return parent[x]

def unite(x, y):
	if find(x) != find(y):
		if rank[x] > rank[y]:
			parent[y] = x
		else:
			parent[x] = y
		if rank[x] == rank[y]:
			rank[x] = rank[x] + 1

def kruskal(k, n):
	ed = []
	for x in edges:
		lst = list(x)
		lst[0] = lst[1] - k*lst[2]
		ed.append(lst)
		#x[0] = x[1] - k*x[2]

	if len(parent) == 0:
		for x in xrange(n):
			parent.append(x)
			rank.append(0)
	else:
		for x in xrange(n):
			parent[x] = x
			rank[x] = 0

	ed.sort(reverse=True)

	sumDenom = 0
	sumNum = 0
	resp = 0

	for e in ed:
		if find(e[3]) != find(e[4]):
			unite(e[3], e[4])
			sumNum = sumNum + e[1]
			sumDenom = sumDenom + e[2]
			resp = resp + e[0]

	return (resp, sumNum, sumDenom)

def reduce(i, j):
	div = gcd(i, j)
	if div <= 1:
		return (i, j)
	return (i/div, j/div)

def binSearch(n):
	lo = 0.0
	hi = float(1e7)
	while hi - lo > 1e-8:
		mid = lo + (hi-lo)/2.0
		a, b, c = kruskal(mid, n)
		if a < 0:
			hi = mid
		else:
			lo = mid
	out = reduce(b, c)
	print("{}/{}".format(out[0], out[1]))


n, m = map(int, raw_input().split())
for x in xrange(m):
	x,y,z,w = map(int, raw_input().split())
	edges.append((float(z)/w, z, w, x, y))
binSearch(n)

File: HackerRank/stf.cpp
/*input
3 3
0 1 1 1
1 2 2 4
2 0 1 2
*/
#include <bits/stdc++.h>

using namespace std;

#define fi first
#define se second
#define all(v) (v).begin(), (v).end()
#define rall(v) (v).rbegin(), (v).rend()
#define mt make_tuple
#define mp make_pair

typedef vector<int> vi;
// 			 a - k*x, a,  b,   s,   t
vector<tuple<double, int, int, int, int> > edges;

const int maxn = 1e5 + 10;
const double tol = 1e-7;

int n, m;
vector<int> parent(maxn);
vector<int> rk(maxn, 0);



int find(int x)
{
	if(x != parent[x])
		parent[x] = find(parent[x]);
	return parent[x];
}

void unite(int a, int b)
{
	if(find(a) != find(b))
	{
		if(rk[a] > rk[b]) parent[b] = a;
		else parent[a] = b;	
		if(rk[a] == rk[b]) rk[a]++;
	}
}

tuple<long double, int, int> kruskal(double K)
{
	iota(all(parent), 0);
	fill(all(rk), 0);
	for(auto& p : edges){
		get<0>(p) = (get<1>(p) - K * get<2>(p));	
	} 
	sort(rall(edges));
	double sum = 0;
	int numSum = 0, denomSum = 0;
	for(auto& p : edges)
	{
		if(find(get<3>(p)) != find(get<4>(p)) )
		{
			numSum += get<1>(p);
			denomSum += get<2>(p);
			unite(get<3>(p), get<4>(p));
			sum += get<0>(p);
		}
	}
	return mt(sum, numSum, denomSum);
}

pair<int, int> irreducible(int a, int b)
{
	int d = __gcd(a, b);
	return (d == 1) ? mp(a, b) : mp(a/d, b/d);
}

pair<int, int> binSearch()
{
	double beg = 0, end = 20000010, value = 10000;
	int numSum = 0, denomSum = 0;
	while(end - beg > tol)
	{
		double mid = (beg + end)/2;
		tie(value, numSum, denomSum) = kruskal(mid);
		if(value > 0)
			beg = mid;
		else
			end = mid;
	}
	return irreducible(numSum, denomSum);
}

int main()
{
	int s, t, x, y;
	ios::sync_with_stdio(0); cin.tie(0);
	cin >> n >> m;
	for(int i = 0; i < m; ++i)
	{
		cin >> s >> t >> x >> y;
		edges.push_back(mt((double)x/(double)y, x, y, s, t));
	}	
	auto resp = binSearch();
	cout << resp.fi << "/" << resp.se << endl;
	return 0;
}